server/src/main/java/org/elasticsearch/action/admin/cluster/configuration/TransportClearVotingTombstonesAction.java

```diff
@@ -0,0 +1,150 @@
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ElasticsearchTimeoutException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.master.TransportMasterNodeAction;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterState.Builder;
+import org.elasticsearch.cluster.ClusterStateObserver;
+import org.elasticsearch.cluster.ClusterStateObserver.Listener;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.function.Predicate;
+
+public class TransportClearVotingTombstonesAction
+    extends TransportMasterNodeAction<ClearVotingTombstonesRequest, ClearVotingTombstonesResponse> {
+
+    @Inject
+    public TransportClearVotingTombstonesAction(TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+                                                ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(ClearVotingTombstonesAction.NAME, transportService, clusterService, threadPool, actionFilters,
+            ClearVotingTombstonesRequest::new, indexNameExpressionResolver);
+    }
+
+    @Override
+    protected String executor() {
+        return Names.SAME;
+    }
+
+    @Override
+    protected ClearVotingTombstonesResponse newResponse() {
+        throw new UnsupportedOperationException("usage of Streamable is to be replaced by Writeable");
+    }
+
+    @Override
+    protected ClearVotingTombstonesResponse read(StreamInput in) throws IOException {
+        return new ClearVotingTombstonesResponse(in);
+    }
+
+    @Override
+    protected void masterOperation(ClearVotingTombstonesRequest request, ClusterState initialState,
+                                   ActionListener<ClearVotingTombstonesResponse> listener) throws Exception {
+
+        final long startTimeMillis = threadPool.relativeTimeInMillis();
+
+        final Predicate<ClusterState> allTombstonedNodesRemoved = new Predicate<ClusterState>() {
+            @Override
+            public boolean test(ClusterState newState) {
+                for (DiscoveryNode tombstone : initialState.getVotingTombstones()) {
+                    if (newState.nodes().nodeExists(tombstone.getId())) {
```

Comment:
add a comment here saying that you deliberately don't want to implement `nodeExists(DiscoveryNode)` because that checks for equality of ephemeral id.

Reply:
Done.